File: backend/src/mocks/handler/startMockHandler.ts
import { Request, Response } from 'express';
import { spawn } from 'child_process';
import fs from 'fs';
import path from 'path';
import * as console from 'node:console';
import { isPortInUse, isPortSafe } from '@/utils/portUtils';
import { ApiResponse, StartMockResponse } from '@/types';
import { fileRepository } from '@/mocks/repositories/fileRepository';
import { mockInstanceRepository } from '@/mocks/repositories/mockInstanceRepository';
import { LOGS_DIR } from '@/lib/constants';
import { prisma } from '@/prisma';
import { generateDynamicTraefikConfig } from '@/traefik/generate-traefik-config';

const useUnsafePort = true;

export async function startMockHandler(req: Request, res: Response<ApiResponse<StartMockResponse>>) {
	const { port, configFile, uuid } = req.body;
	console.log('Starting mock server with config:', configFile, 'on port:', port);
	try {
		if (port && !isPortSafe(port) && !useUnsafePort) {
			return res.status(400).json({
				error: 'Invalid port. Port must be between 9001 and 9999.'
			});
		}

		const portInUse = await isPortInUse(port);
		if (portInUse) {
			return res.status(400).json({
				error: `Port ${port} is already in use.`
			});
		}

		const configPath = await fileRepository.getConfigPath(configFile);
		if (!await fileRepository.configExists(configFile)) {
			return res.status(404).json({ error: 'Configuration file not found' });
		}

		const mockProcess = spawn('mockoon-cli', [
			'start',
			'--data',
			configPath,
			'--port',
			port.toString()
		]);

		const logFile = fs.createWriteStream(
			path.join(LOGS_DIR, `mock-${port}.log`),
			{ flags: 'a' }
		);

		mockProcess.stdout.pipe(logFile);
		mockProcess.stderr.pipe(logFile);

		mockInstanceRepository.add(port, {
			process: mockProcess,
			configFile,
			startTime: new Date(),
			logFile,
			uuid
		});

		let alias = await prisma.alias.findFirst({
			where: {
				fileName: configFile
			}
		})

		if (!alias) {
			alias = await prisma.alias.create({
				data: {
					fileName: configFile,
					alias: port.toString(),
					port: port,
					isActive: true
				}
			}).catch(e=>{
				console.error('Error creating alias:', e);
				return null;
			})
		}else{
			alias = await prisma.alias.update({
				where: {
					id: alias.id
				},
				data: {
					port: port,
					isActive: true
				}
			}).catch(e=>{
				console.error('Error updating alias:', e);
				return null;
			})
		}

		if (!alias) {
			return res.status(500).json({ error: 'Failed to create alias' });
		}

		await generateDynamicTraefikConfig()
			.catch(e => {
				console.error('Error generating dynamic Traefik config:', e);
			});

		const response: StartMockResponse = {
			success: true,
			uuid,
			port,
			configFile,
			message: `Mock server started on port ${port}`
		};

		res.json(response);
	} catch (error: any) {
		console.error('Error starting mock server:', error);
		res.status(500).json({ error: error.message });
	}
}

File: backend_new/src/mocks/handler/startMockHandler.go
package handler

import (
	"log"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"time"

	"github.com/gin-gonic/gin"

	"mockoon-control-panel/backend_new/src/lib"
	"mockoon-control-panel/backend_new/src/mocks/repositories"
	"mockoon-control-panel/backend_new/src/traefik"
	"mockoon-control-panel/backend_new/src/types"
	"mockoon-control-panel/backend_new/src/utils"
)

// UseUnsafePort determines if we should allow ports outside the safe range
const UseUnsafePort = true

// StartMockHandler handles requests to start a mock server
func StartMockHandler(c *gin.Context) {
	var req types.StartMockRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request format: " + err.Error()})
		return
	}

	log.Printf("Starting mock server with config: %s on port: %d", req.ConfigFile, req.Port)

	// Validate port if not using unsafe port
	if !UseUnsafePort && !utils.IsPortSafe(req.Port) {
		c.JSON(http.StatusBadRequest, gin.H{
			"error": "Invalid port. Port must be between 9001 and 9999.",
		})
		return
	}

	// Check if port is already in use
	if utils.IsPortInUse(req.Port) {
		c.JSON(http.StatusBadRequest, gin.H{
			"error": "Port " + strconv.Itoa(req.Port) + " is already in use.",
		})
		return
	}

	// Check if config file exists
	configPath := repositories.FileRepo.GetConfigPath(req.ConfigFile)
	if !repositories.FileRepo.ConfigExists(req.ConfigFile) {
		c.JSON(http.StatusNotFound, gin.H{
			"error": "Configuration file not found",
		})
		return
	}

	// Create log directory if it doesn't exist
	if err := utils.EnsureDirectoryExists(lib.LOGS_DIR); err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"error": "Failed to create log directory: " + err.Error(),
		})
		return
	}

	// Create log file
	logFilePath := filepath.Join(lib.LOGS_DIR, "mock-"+strconv.Itoa(req.Port)+".log")
	logFile, err := os.OpenFile(logFilePath, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0644)
	if err != nil {
		c.JSON(http.StatusInternalServerError, gin.H{
			"error": "Failed to create log file: " + err.Error(),
		})
		return
	}

	// Start mockoon-cli process
	cmd := exec.Command("mockoon-cli",
		"start",
		"--data", configPath,
		"--port", strconv.Itoa(req.Port),
	)

	// Redirect stdout and stderr to log file
	cmd.Stdout = logFile
	cmd.Stderr = logFile

	// Start the process
	if err := cmd.Start(); err != nil {
		logFile.Close()
		c.JSON(http.StatusInternalServerError, gin.H{
			"error": "Failed to start mock server: " + err.Error(),
		})
		return
	}

	// Record the instance
	now := time.Now()
	repositories.MockInstanceRepo.Add(req.Port, types.MockInstance{
		Process:    cmd.Process,
		Cmd:        cmd,
		ConfigFile: req.ConfigFile,
		Port:       req.Port,
		StartTime:  now,
		LogFile:    logFile,
		UUID:       req.UUID,
	})

	// Update Traefik configuration
	if err := traefik.GenerateDynamicTraefikConfig(); err != nil {
		log.Printf("Warning: Failed to update Traefik configuration: %v", err)
	}

	// Return success response
	c.JSON(http.StatusOK, gin.H{
		"data": types.StartMockResponse{
			Port:       req.Port,
			ConfigFile: req.ConfigFile,
			StartTime:  now,
			Status:     "running",
		},
	})
}
